src/main.rs

```diff
@@ -3,43 +3,129 @@ mod error;
-use replacer::Replacer;
+pub(crate) use self::input::Source;
+pub(crate) use self::error::{Error, Result};
+use self::replacer::Replacer;
+use self::input::{make_mmap, make_mmap_stdin};
 
-use clap::Parser;
+fn main() -> Result<()> {
+    let options = cli::Options::parse();
+
+    let replacer = Replacer::new(
+        options.find,
+        options.replace_with,
+        options.literal_mode,
+        options.flags,
+        options.replacements,
+    )?;
+
+    let sources = if !options.files.is_empty() {
+        Source::from_paths(options.files)
+    } else {
+        Source::from_stdin()
+    };
+
+    let mut errors = Vec::new();
+    
+    let mut mmaps = Vec::new();
+    for source in sources.iter() {
+        let maybe_mmap = match source {
+            Source::File(path) => if path.exists() {
+                make_mmap(&path)
+            } else {
+                Err(Error::InvalidPath(path.clone()))
+            },
+            Source::Stdin => make_mmap_stdin()
+        };
+
+        match maybe_mmap {
+            Ok(mmap) => mmaps.push(Some(mmap)),
+            Err(e) => {
+                mmaps.push(None);
+                errors.push((source, e));
+            },
+        };
+    }
+
+    let needs_separator = sources.len() > 1;
+
+    let replaced: Vec<_> = {
+        use rayon::prelude::*;
+        mmaps.par_iter()
+            .map(|maybe_mmap| (maybe_mmap.as_ref().map(|mmap| replacer.replace(&mmap))))
+            .collect()
+    };
+
+    if options.preview || sources.first() == Some(&Source::Stdin) {
+        let mut handle = stdout().lock();
 
-fn main() {
-    if let Err(e) = try_main() {
-        eprintln!("{}: {}", Style::from(Color::Red).bold().paint("error"), e);
+        for (source, replaced) in sources.iter().zip(replaced).filter(|(_, r)| r.is_some()) {
+            if needs_separator {
+                writeln!(handle, "----- {} -----", source.display())?;
+            }
+            handle.write_all(replaced.as_ref().unwrap())?;
+        }
+    } else {
+        // Windows requires closing mmap before writing:
+        // > The requested operation cannot be performed on a file with a user-mapped section open
+        #[cfg(target_family = "windows")]
+        let replaced: Vec<Option<Vec<u8>>> = replaced.into_iter().map(|r| r.map(|r| r.to_vec())).collect();
+        #[cfg(target_family = "windows")]
+        drop(mmaps);
+
+        for (source, replaced) in sources.iter().zip(replaced).filter(|(_, r)| r.is_some()) {
+            let path = match source {
+                Source::File(path) => path,
+                _ => unreachable!("stdin should go previous branch"),
+            };
+            if let Err(e) = write_with_temp(path, &replaced.unwrap()) {
+                errors.push((source, e));
+            }
+        }
+    }
+
+    if !errors.is_empty() {
+        for (source, error) in errors {
+            eprintln!("error: {}: {:?}", source.display(), error);
+        }
         process::exit(1);
     }
```

Comment:
> What's "that variant for failed jobs" exactly? `Error::FailedJobs`? They were collected from `Replacer::replace_file` runs, which were replaced with match arms.

Yes, I'm talking about the removal of `FailedJobs`. This loop here is serving the same purpose, but this is splitting how errors are displayed to the user where all of the other errors are bubbled up out of `main()` aside from this one

Reply:
There are only 2 errors "bubbling up out of `main()`":

https://github.com/nc7s/sd/blob/601db39cae34b3a3a6a22bc393e49fe724de0cbb/src/main.rs#L75-L77

… both are writes to stdout, which I think is fine to skip "job" error handling in the loop. I actually wanted to clear them out using `.expect()` or something, but didn't think of a good way without panicking.

Reply:
There is also the capture group name validation error which is why the error message snapshots all had their text changed from `error` to `Error`

I'm not really wanting to have a long-winded argument here. If you want we can merge this as-is and I can make changes to the error handling code to handle collecting and presenting multiple errors for some of the variants. I've been holding off on making changes till after this PR is merged, but that's taken longer than expected

Reply:
That's overlooked. I'm not insisting on that, but didn't see what "bubbles" were. Code might convey the meanings better.

Reply:
Apologies. I agree code would be the best way to express things. I'll throw up a branch after I'm off work to show what I mean

"Bubbling up" is a common phrase for error handling probably mostly influenced by other languages where errors are automatically "bubbled up" till they're explicitly caught. For Rust it's just a more confusing way to say returning an error I suppose 😅

Reply:
Finally got around to handling this. Here's the branch if you want to cherry-pick commits  or anything

https://github.com/CosmicHorrorDev/sd/commits/refactor-review

The gist of the changes that I was failing to communicate before is that you can fail fast when possible, but once replacements start being made then you pass the "point of no return" and just collect all errors and make any replacements that you can. We do this because we don't want to leave a partially operated on state when we're already making replacements, but if we can report a failure before making any cases then that's preferable

Reply:
That's fair, totally agreed. I have no further thoughts here.

Reply:
Great! Well feel free to pull over my changes if they look good to you and we can get this PR merged 🎉

Reply:
Sure. IIRC you can also push to this PR, but that's more work for you, just saying if you wanna try it next time ;)